File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/access/SelectedBrowserLibraryProvider.java
package com.lasthopesoftware.bluewater.client.library.access;

import com.lasthopesoftware.bluewater.client.library.repository.Library;
import com.lasthopesoftware.bluewater.client.servers.selection.ISelectedLibraryIdentifierProvider;
import com.lasthopesoftware.messenger.promises.Promise;

/**
 * Created by david on 2/12/17.
 */

public class SelectedBrowserLibraryProvider implements ISelectedBrowserLibraryProvider {

	private final ISelectedLibraryIdentifierProvider selectedLibraryIdentifierProvider;
	private final ILibraryProvider libraryProvider;

	public SelectedBrowserLibraryProvider(ISelectedLibraryIdentifierProvider selectedLibraryIdentifierProvider, ILibraryProvider libraryProvider) {
		this.selectedLibraryIdentifierProvider = selectedLibraryIdentifierProvider;
		this.libraryProvider = libraryProvider;
	}

	@Override
	public Promise<Library> getBrowserLibrary() {
		return libraryProvider.getLibrary(selectedLibraryIdentifierProvider.getSelectedLibraryId());
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/library/access/session/SelectedBrowserLibraryProvider.kt
package com.lasthopesoftware.bluewater.client.browsing.library.access.session

import com.lasthopesoftware.bluewater.client.browsing.library.access.ILibraryProvider
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library
import com.lasthopesoftware.bluewater.shared.promises.extensions.keepPromise
import com.namehillsoftware.handoff.promises.Promise

class SelectedBrowserLibraryProvider(private val selectedLibraryIdentifierProvider: ProvideSelectedLibraryId, private val libraryProvider: ILibraryProvider) : ISelectedBrowserLibraryProvider {
	override fun getBrowserLibrary(): Promise<Library?> =
		selectedLibraryIdentifierProvider.selectedLibraryId.eventually {
			it?.let(libraryProvider::getLibrary).keepPromise()
		}
}
